### PR TITLE
VB: Unify CreateCompilation helpers

### PR DESCRIPTION
**Pull request by jaredpar**
Next phase in the VB test helper changes. This unifies our CreateCompilation helpers to match the style of the C# counterparts. 

Items specifically excluded from this change:

- Changes to CreateCompilation...AndReferences. These methods will be going away as they are redundant with their counterparts that lack the AndReferences suffix. 
- Changes to CompileAndVerify 

Both of these items are coming but they are largely refactoring operations. Going to separate that out into a separate PR to help with the reviews.

### CODE REVIEW

**Comment by jaredpar** [comment]
CC @dotnet/roslyn-compiler for review

File: NuGet.Config [hunk 1]
```diff
@@ -29,5 +29,6 @@
     <add key="dotnet.myget.org roslyn" value="https://dotnet.myget.org/F/roslyn/api/v3/index.json" />
     <add key="myget.org vs-editor" value="https://myget.org/F/vs-editor/api/v3/index.json" />
     <add key="nuget.org" value="https://api.nuget.org/v3/index.json" />
+    <add key="DO NOT MERGE" value="E:\code\roslyn-internal\Closed\Test\Resources\bin\Debug" />
```
Comment (jcouv) [inline NuGet.Config]: >DO NOT MERGE [](start = 14, length = 12)

This means "do not merge to master" or "do not merge at all"?

File: src/Compilers/Test/Utilities/VisualBasic/BasicTestBase.vb [hunk 2]
```diff
@@ -186,7 +186,7 @@ Public MustInherit Class BasicTestBase
 
         Dim assemblyName As String = Nothing
         Dim sourceTrees = ParseSourceXml(source, parseOptions, assemblyName)
-        Dim compilation = CreateEmptyCompilation(sourceTrees, allReferences, options, assemblyName)
+        Dim compilation = CreateEmptyCompilation(sourceTrees.ToArray(), allReferences, options, assemblyName:=assemblyName)
```
Comment (jcouv) [inline src/Compilers/Test/Utilities/VisualBasic/BasicTestBase.vb]: Nit: `assemblyName:=` probably can be removed.

File: src/Compilers/Test/Utilities/VisualBasic/BasicTestBase.vb [hunk 3]
```diff
@@ -385,6 +384,16 @@ Public MustInherit Class BasicTestBase
...
     End Function
 
+    Friend Overloads Function CompileAndVerifyFieldMarshal(source As String,
+                                                           expectedBlobs As Dictionary(Of String, Byte()),
+                                                           Optional getExpectedBlob As Func(Of String, PEAssembly, Byte()) = Nothing,
+                                                           Optional expectedSignatures As SignatureDescription() = Nothing,
+                                                           Optional isField As Boolean = True) As CompilationVerifier
+        Dim xmlSource = <compilation><field><%= source %></field></compilation>
+        Return CompileAndVerifyFieldMarshal(xmlSource, expectedBlobs, getExpectedBlob, expectedSignatures, isField)
+    End Function
```
Comment (jcouv) [inline src/Compilers/Test/Utilities/VisualBasic/BasicTestBase.vb]: Nit: extra empty line below

File: NuGet.Config (hunk 1 above)
Comment (jaredpar) [reply NuGet.Config]: That was a reminder to myself to delete this line. I failed.

**Comment by jcouv** [comment]
Sounds like you're going to refresh this PR. Ping me afterwards to resume the review. Thanks

**Comment by jaredpar** [comment]
@jcouv this is ready for review now.

**Comment by jaredpar** [comment]
CC @roslyn-compiler for review. The change is actually focused in the following files:

- BasicTestSource.vb
- BasicTestBase.vb
- CompilationTestUtils.vb
- CommonTestBase.cs
- TargetFrameworkUtil.cs

Everything else is just an API change reacting to these changes.

File: src/Compilers/Test/Utilities/VisualBasic/BasicTestBase.vb (hunk 2 above)
Comment (cston) [inline src/Compilers/Test/Utilities/VisualBasic/BasicTestBase.vb]: >.ToArray() [](start = 60, length = 10)

Is `.ToArray()` necessary here?

File: src/Compilers/Test/Utilities/VisualBasic/BasicTestBase.vb [hunk 4]
```diff
@@ -810,8 +799,13 @@ Public MustInherit Class BasicTestBase
 
         Dim fileName = "a.vb"
         Dim syntaxTree = Parse(source, fileName, parseOptions)
-        Dim defaultRefs = If(useLatestFrameworkReferences, LatestVbReferences, DefaultVbReferences)
-        Dim compilation = CreateCompilationWithMscorlib45AndVBRuntime({syntaxTree}, references:=defaultRefs.Append({ValueTupleRef, SystemRuntimeFacadeRef}), options:=If(compilationOptions, TestOptions.ReleaseDll))
+        Dim allReferences As IEnumerable(Of MetadataReference)
+        If useLatestFrameworkReferences Then
```
Comment (cston) [inline src/Compilers/Test/Utilities/VisualBasic/BasicTestBase.vb]: >useLatestFrameworkReferences [](start = 11, length = 28)

Consider moving `useLatestFrameworkReferences` check into `.Add()` call.

File: src/Compilers/Test/Utilities/VisualBasic/BasicTestBase.vb [hunk 5]
```diff
@@ -869,10 +863,15 @@ Public MustInherit Class BasicTestBase
 
         Dim fileName = "a.vb"
         Dim syntaxTree = Parse(source, fileName, parseOptions)
-        Dim defaultRefs = If(useLatestFramework, LatestVbReferences, DefaultVbReferences)
-        Dim allReferences = defaultRefs.Concat({ValueTupleRef, SystemRuntimeFacadeRef})
+        Dim allReferences As IEnumerable(Of MetadataReference) = Nothing
+        If useLatestFramework Then
```
Comment (cston) [inline src/Compilers/Test/Utilities/VisualBasic/BasicTestBase.vb]: >useLatestFramework [](start = 11, length = 18)

Consider moving into `.Add()` call.

File: src/Compilers/Test/Utilities/VisualBasic/CompilationTestUtils.vb [hunk 6]
```diff
@@ -216,7 +158,7 @@ Friend Module CompilationUtils
                                                               Optional assemblyName As String = Nothing) As VisualBasicCompilation
 
         Dim references = {MscorlibRef, SystemRef, MsvbRef}
-        Return CreateEmptyCompilation(source, references, options, assemblyName)
+        Return CreateEmptyCompilation(source.ToArray(), references, options:=options, assemblyName:=assemblyName)
```
Comment (cston) [inline src/Compilers/Test/Utilities/VisualBasic/CompilationTestUtils.vb]: >.ToArray() [](start = 44, length = 10)

Is `.ToArray()` necessary?

File: src/Compilers/Test/Utilities/VisualBasic/BasicTestBase.vb (hunk 2 above)
Comment (jaredpar) [reply src/Compilers/Test/Utilities/VisualBasic/BasicTestBase.vb]: Yes. There is no way to do an implicit conversion from `IEnumerable(Of SyntaxTree)` to `BasicTestSource` as we don't allow widening conversions on interfaces. 

---
In reply to: [192156452](https://github.com/dotnet/roslyn/pull/25130#discussion_r192156452) [](ancestors = 192156452)

File: src/Compilers/Test/Utilities/VisualBasic/BasicTestBase.vb (hunk 4 above)
Comment (jaredpar) [reply src/Compilers/Test/Utilities/VisualBasic/BasicTestBase.vb]: Will do on the next PR here. 

---
In reply to: [192157327](https://github.com/dotnet/roslyn/pull/25130#discussion_r192157327) [](ancestors = 192157327)

File: src/Compilers/Test/Utilities/VisualBasic/BasicTestBase.vb (hunk 5 above)
Comment (jaredpar) [reply src/Compilers/Test/Utilities/VisualBasic/BasicTestBase.vb]: Will do on the next PR here.

---
In reply to: [192157523](https://github.com/dotnet/roslyn/pull/25130#discussion_r192157523) [](ancestors = 192157523)

**Pull request merged**